core/src/workflow/workflow_tasks/mod.rs

```diff
@@ -840,10 +840,11 @@ impl WorkflowTaskManager {
     ///
     /// Any subsequent action that needs to be taken will be created as a new activation
     fn on_activation_done(&self, run_id: &str) {
-        if self.workflow_machines.delete_activation(run_id).is_some() {
-            self.pending_activations_notifier.notify_waiters();
-        }
-        // It's possible the activation is already removed due to completing an eviction
+        self.workflow_machines.delete_activation(run_id);
+        // It's important to use `notify_one` here to avoid possible races where we're waiting
+        // on a cache slot and fail to realize pending activations must be issued before a slot
+        // will free up.
+        self.pending_activations_notifier.notify_one();
```

Comment:
This is the actual fix. When this was protected by the if / not storing the notification, if timing was juuuust right we could be sitting waiting on the cache when we should've realized we instead need to issue a pending activation in order to keep making progress.